pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -166,10 +194,34 @@ func (h *upgradeHandler) OnChanged(key string, upgrade *harvesterv1.Upgrade) (*h
 		toUpdate.Status.SingleNode = singleNode
 
 		repoInfo, err := repo.getInfo()
+		// retry repo.getInfo by requeueing
 		if err != nil {
-			setUpgradeCompletedCondition(toUpdate, StateFailed, corev1.ConditionFalse, err.Error(), "")
-			return h.upgradeClient.Update(toUpdate)
+			retryCount := 0
+			if toUpdate.Annotations == nil {
+				toUpdate.Annotations = make(map[string]string)
+				toUpdate.Annotations[getRepoRetryCountAnnotation] = "0"
+			}
+
+			retryCount, err = strconv.Atoi(toUpdate.Annotations[getRepoRetryCountAnnotation])
```

Comment:
If any controller set annotations, but there is no `getRepoRetryCountAnnotation`. We may have an error here. We can update like:

```suggestion
        retryCountStr, ok := toUpdate.Annotations[getRepoRetryCountAnnotation]
        if !ok {
            retryCountStr = "0"
        }
        retryCount, err = strconv.Atoi(retryCountStr)
```

Reply:
This one is obsolete. The relevant commit is cherry-picked and merged with a different solution in #3486. I will rebase and push again. Thanks!